worker/src/RTC/IceServer.cpp

```diff
@@ -98,296 +131,383 @@ namespace RTC
+	}
+
+	void IceServer::RemoveTuple(RTC::TransportTuple* tuple)
+	{
+		MS_TRACE();
+
+		RTC::TransportTuple* removedTuple{ nullptr };
+
+		// Find the removed tuple.
+		auto it = this->tuples.begin();
+
+		for (; it != this->tuples.end(); ++it)
+		{
+			RTC::TransportTuple* storedTuple = std::addressof(*it);
+
+			if (storedTuple->Compare(tuple))
 			{
-				MS_WARN_TAG(ice, "ignoring STUN Binding Response without FINGERPRINT");
+				removedTuple = storedTuple;
+
+				break;
 			}
+		}
 
+		// If not found, ignore.
+		if (!removedTuple)
+		{
 			return;
 		}
 
-		switch (packet->GetClass())
-		{
-			case RTC::StunPacket::Class::REQUEST:
-			{
-				// USERNAME, MESSAGE-INTEGRITY and PRIORITY are required.
-				if (!packet->HasMessageIntegrity() || (packet->GetPriority() == 0u) || packet->GetUsername().empty())
-				{
-					MS_WARN_TAG(ice, "mising required attributes in STUN Binding Request => 400");
+		// Notify the listener.
+		this->listener->OnIceServerTupleRemoved(this, removedTuple);
 
-					// Reply 400.
-					RTC::StunPacket* response = packet->CreateErrorResponse(400);
+		// Remove it from the list of tuples.
+		// NOTE: Do it after notifying the listener since the listener may need to
+		// use/read the tuple being removed so we cannot free it yet.
+		this->tuples.erase(it);
 
-					response->Serialize(StunSerializeBuffer);
-					this->listener->OnIceServerSendStunPacket(this, response, tuple);
+		// If this is the selected tuple, do things.
+		if (removedTuple == this->selectedTuple)
+		{
+			this->selectedTuple = nullptr;
 
-					delete response;
+			// Mark the first tuple as selected tuple (if any) but only if state was
+			// 'connected' or 'completed'.
+			if (
+			  (this->state == IceState::CONNECTED || this->state == IceState::COMPLETED) &&
+			  this->tuples.begin() != this->tuples.end())
+			{
+				SetSelectedTuple(std::addressof(*this->tuples.begin()));
 
-					return;
+				// Restart ICE consent check to let the client send new consent requests
+				// on the new selected tuple.
+				if (IsConsentCheckSupported())
+				{
+					RestartConsentCheck();
 				}
+			}
+			// Or just emit 'disconnected'.
+			else
+			{
+				// Update state.
+				this->state = IceState::DISCONNECTED;
 
-				// Check authentication.
-				switch (packet->CheckAuthentication(this->usernameFragment, this->password))
-				{
-					case RTC::StunPacket::Authentication::OK:
-					{
-						if (!this->oldUsernameFragment.empty() && !this->oldPassword.empty())
-						{
-							MS_DEBUG_TAG(ice, "new ICE credentials applied");
+				// Reset remote nomination.
+				this->remoteNomination = 0u;
 
-							// Notify the listener.
-							this->listener->OnIceServerLocalUsernameFragmentRemoved(this, this->oldUsernameFragment);
+				// Notify the listener.
+				this->listener->OnIceServerDisconnected(this);
 
-							this->oldUsernameFragment.clear();
-							this->oldPassword.clear();
-						}
+				if (IsConsentCheckSupported() && IsConsentCheckRunning())
```

Comment:
This would suffice:

```suggestion
				if (IsConsentCheckRunning())
```

Reply:
Yes, I know. But I'd prefer to be super explicit because lot of logic in IceServer file is too sensitive and probably should be refactored in the future.

Reply:
Rationale is that I want to prevent that consent check is started when not supported. It cannot happen because the `StartConsentCheck()` method asserts `IsConsentCheckSupported()` but anyway.